flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -760,33 +763,26 @@ private void doRun() {
             // make sure the user code classloader is accessible thread-locally
             executingThread.setContextClassLoader(userCodeClassLoader.asClassLoader());
 
-            FlinkSecurityManager.monitorUserSystemExitForCurrentThread();
+            TaskInvokable finalInvokable = invokable;
             try {
-                // Restore invokable data to the last valid state
-                invokable.restore();
-            } finally {
-                FlinkSecurityManager.unmonitorUserSystemExitForCurrentThread();
-            }
+                runWithSystemExitMonitoring(finalInvokable::restore);
 
-            if (!transitionState(ExecutionState.INITIALIZING, ExecutionState.RUNNING)) {
-                throw new CancelTaskException();
-            }
+                if (!transitionState(ExecutionState.INITIALIZING, ExecutionState.RUNNING)) {
+                    throw new CancelTaskException();
+                }
 
-            // notify everyone that we switched to running
-            taskManagerActions.updateTaskExecutionState(
-                    new TaskExecutionState(executionId, ExecutionState.RUNNING));
+                // notify everyone that we switched to running
+                taskManagerActions.updateTaskExecutionState(
+                        new TaskExecutionState(executionId, ExecutionState.RUNNING));
 
-            // Monitor user codes from exiting JVM covering user function invocation. This can be
-            // done in a finer-grained way like enclosing user callback functions individually,
-            // but as exit triggered by framework is not performed and expected in this invoke
-            // function anyhow, we can monitor exiting JVM for entire scope.
-            FlinkSecurityManager.monitorUserSystemExitForCurrentThread();
-            try {
-                // run the invokable
-                invokable.invoke();
-            } finally {
-                FlinkSecurityManager.unmonitorUserSystemExitForCurrentThread();
+                runWithSystemExitMonitoring(finalInvokable::invoke);
+            } catch (Throwable throwable) {
+                if (!(throwable instanceof CancelTaskException)) {
```

Comment:
I don't sure about this condition. According to this code if CancelTaskException happens we just ignore it and continue the execution but the old implementation always throws exceptions.
Right now, the logic is pretty confusing - if the problem happens inside of `restore` or `invoke` we cancel the task immediately in catch block but if it happens between these two calls(during the state transit) we will cancel the task a little later in the different catch block. I think it won't be wrong to unify this logic and cancel the task here whatever exception happened. WDYT?

Reply:
It looks strange to me as well.

Reply:
Good points, I addressed this in 6843a959b16a601a3db3247eaa77e46e05b0f900 and 4d64111967e37e0333e0b7310eb1108d6c04679a.

So `cancelTask` is called unconditionally now from `cleanUp`.
(I still left a separate `invokable.cancel()` call in Task when transitioning state or cancelling externally because even if we unify it for `StreamTask`, other Invokables can have diffferent logic in `cancel`).

Reply:
as I understand, right now we call `cancelTask` always regardless we have an exception or not. I am ok with that but I only want to emphasize that this behavior is different from the old one when `cancelTask` was used only for an exception case. So if tests are green and nobody sees the problem with that we can leave it likes that.

Reply:
I have adjusted the logic so that `cancelTask()` is called from `cleanUp()` only in exceptional case.

Otherwise, `SourceFunction` gets cancelled in the normal case (which is probably fine, but FLINK-23527 should be  resolved first).